Git-commit: 8c272261194dfda11cc046fbe808e052f6f284eb
From: Nishanth Aravamudan <nacc@linux.vnet.ibm.com>
Date: Mon, 19 May 2014 11:14:23 -0700
Subject: [PATCH] powerpc/numa: Enable USE_PERCPU_NUMA_NODE_ID
Patch-mainline: v3.16-rc1
Reference: bsc#924809

Based off 3bccd996 for ia64, convert powerpc to use the generic per-CPU
topology tracking, specifically:

    initialize per cpu numa_node entry in start_secondary
    remove the powerpc cpu_to_node()
    define CONFIG_USE_PERCPU_NUMA_NODE_ID if NUMA

Signed-off-by: Nishanth Aravamudan <nacc@linux.vnet.ibm.com>
Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Acked-by: Dinar Valeev <dvaleev@suse.com>
---
 arch/powerpc/Kconfig                |  4 ++++
 arch/powerpc/include/asm/topology.h | 13 -------------
 arch/powerpc/kernel/smp.c           |  6 ++++++
 3 files changed, 10 insertions(+), 13 deletions(-)

--- dev1/arch/powerpc/Kconfig	2015-03-30 18:13:46.360069816 -0500
+++ dev1/arch/powerpc/Kconfig	2015-03-30 18:42:39.960098779 -0500
@@ -447,6 +447,10 @@ config NODES_SHIFT
 	default "4"
 	depends on NEED_MULTIPLE_NODES
 
+config USE_PERCPU_NUMA_NODE_ID
+	def_bool y
+	depends on NUMA
+
 config ARCH_SELECT_MEMORY_MODEL
 	def_bool y
 	depends on PPC64
--- dev1/arch/powerpc/include/asm/topology.h	2015-03-30 18:13:48.760069817 -0500
+++ dev1/arch/powerpc/include/asm/topology.h	2015-03-30 18:42:39.980097358 -0500
@@ -16,19 +16,6 @@ struct device_node;
 
 #include <asm/mmzone.h>
 
-static inline int cpu_to_node(int cpu)
-{
-	int nid;
-
-	nid = numa_cpu_lookup_table[cpu];
-
-	/*
-	 * During early boot, the numa-cpu lookup table might not have been
-	 * setup for all CPUs yet. In such cases, default to node 0.
-	 */
-	return (nid < 0) ? 0 : nid;
-}
-
 #define parent_node(node)	(node)
 
 #define cpumask_of_node(node) ((node) == -1 ?				\
--- dev1/arch/powerpc/kernel/smp.c	2015-03-30 18:13:47.070069816 -0500
+++ dev1/arch/powerpc/kernel/smp.c	2015-03-30 18:42:39.980097358 -0500
@@ -396,6 +396,7 @@ void smp_prepare_boot_cpu(void)
 #ifdef CONFIG_PPC64
 	paca[boot_cpuid].__current = current;
 #endif
+	set_numa_node(numa_cpu_lookup_table[boot_cpuid]);
 	current_set[boot_cpuid] = task_thread_info(current);
 }
 
@@ -727,6 +728,11 @@ void start_secondary(void *unused)
 	}
 	traverse_core_siblings(cpu, true);
 
+	/*
+	 * numa_node_id() works after this.
+	 */
+	set_numa_node(numa_cpu_lookup_table[cpu]);
+
 	smp_wmb();
 	notify_cpu_starting(cpu);
 	set_cpu_online(cpu, true);
